images/hub/jupyterhub_config.py

```diff
@@ -22,70 +22,63 @@
 c.JupyterHub.cleanup_servers = False
 
 # Check that the proxy has routes appropriately setup
-# This isn't the best named setting :D
 c.JupyterHub.last_activity_interval = 60
 
-# Max number of servers that can be spawning at any one time
-c.JupyterHub.concurrent_spawn_limit = get_config('hub.concurrent-spawn-limit')
-
 # Don't wait at all before redirecting a spawning user to the progress page
 c.JupyterHub.tornado_settings = {
     'slow_spawn_timeout': 0,
 }
 
-# Max number of consecutive failures before the Hub restarts itself
-# requires jupyterhub 0.9.2
-c.Spawner.consecutive_failure_limit = get_config('hub.consecutive-failure-limit', 0)
-
-active_server_limit = get_config('hub.active-server-limit', None)
-if active_server_limit is not None:
-    c.JupyterHub.active_server_limit = int(active_server_limit)
+for trait, cfg_key in (
+    # Max number of servers that can be spawning at any one time
+    ('concurrent_spawn_limit', 'concurrent-spawn-limit'),
+    # Max number of consecutive failures before the Hub restarts itself
+    # requires jupyterhub 0.9.2
+    ('consecutive_failure_limit', 'consecutive-failure-limit'),
+    # Max number of servers to be running at one time
+    ('active_server_limit', 'active-server-limit'),
+):
+    set_config_if_not_none(c.JupyterHub, trait, 'hub.' + cfg_key)
 
 c.JupyterHub.ip = os.environ['PROXY_PUBLIC_SERVICE_HOST']
 c.JupyterHub.port = int(os.environ['PROXY_PUBLIC_SERVICE_PORT'])
 
 # the hub should listen on all interfaces, so the proxy can access it
 c.JupyterHub.hub_ip = '0.0.0.0'
 
-c.KubeSpawner.common_labels = get_config('kubespawner.common-labels')
+set_config_if_not_none(c.KubeSpawner, 'common_labels', 'kubespawner.common-labels')
 
 c.KubeSpawner.namespace = os.environ.get('POD_NAMESPACE', 'default')
 
-c.KubeSpawner.start_timeout = get_config('singleuser.start-timeout')
+set_config_if_not_none(c.KubeSpawner, 'start_timeout', 'singleuser.start-timeout')
 
 # Use env var for this, since we want hub to restart when this changes
 c.KubeSpawner.image_spec = os.environ['SINGLEUSER_IMAGE']
 
-c.KubeSpawner.image_pull_policy = get_config('singleuser.image-pull-policy')
-
-c.KubeSpawner.image_pull_secrets = get_config('singleuser.image-pull-secret-name', None)
-
-c.KubeSpawner.events_enabled = get_config('singleuser.events', False)
-
-c.KubeSpawner.extra_annotations = get_config('singleuser.extra-annotations', {})
-
-c.KubeSpawner.extra_labels = get_config('singleuser.extra-labels', {})
+for trait, cfg_key in (
+    ('start_timeout', 'start-timeout'),
+    ('image_pull_policy', 'image-pull-policy'),
+    ('image_pull_secrets', 'image-pull-secret-name'),
+    ('events_enabled', 'events'),
+    ('extra_labels', 'extra-labels'),
+    ('extra_annotations', 'extra-annotations'),
+    ('uid', 'uid'),
+    ('fs_gid', 'fs-gid'),
```

Comment:
Just a note: We are required by kubespawner to set uid/fs_gid atm I think, perhaps it was image_pull_policy also, a few traitlets are special.

Perhaps we should add some default values on them in kubespawner, allowing values provided to a hub image to still work which yuvi suggested would be good.